src/Console/Output/Progress/ProgressOutputFactory.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Console\Output\Progress;
+
+use PhpCsFixer\Console\Output\OutputContext;
+
+/**
+ * @internal
+ */
+final class ProgressOutputFactory
+{
+    public function create(string $outputType, OutputContext $context): ProgressOutputInterface
+    {
+        if (null === $context->getOutput()) {
+            $outputType = ProgressOutputType::NONE;
```

Comment:
should we raise exception if `getOutput()===null && outputType !== 'none'` ?

Reply:
It's thrown below, if `$outputType` is not any of supported built-in types. This condition here is only about forcing empty output if there's no CLI output available (honestly I don't remember now why it was done like this, it was 3 months ago).

Reply:
I am talking about valid outputTypes. like 'dots'.

especially this test case: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848/files#diff-d9c809453496cbba84934580e854433ee280b478095ad63c62338c51f9c78cb2R52
you silently called back to `none` output type, i wonder about throwing exception of invalid input. would that make sense?

Reply:
Exceptions are for... exceptional situations 😉. If output is no available, we can just fallback to none output, so it makes is non-exceptional IMHO. This condition AFAIR is for some edge cases where `OutputInterface` in the command is not an instance of `ConsoleOutputInterface` and format is other than `txt` - then `$stdErr` is null and then progress output type is overridden to `none`. And that's OK in my opinion. In most scenarios there will be output available.

Reply:
i am basically wondering on which level to make that fallback - here, on factory method, or on class facing user input.

Reply:
IMHO this is the best place for it - user's input is resolved in `ConfigurationResolver` but the output is determined in `FixCommand` based on CLI's I/O and resolved config. Here we have central place that can decide which progress output fits best considering the context.